lib/internal/quic/core.js

```diff
@@ -2577,24 +2597,99 @@ class QuicStream extends Duplex {
-      this.uncork();
-      this.emit('ready');
+      state.sharedState = new QuicStreamSharedState(handle.state);
+      state.session[kAddStream](state.id, this);
     } else {
+      if (this[kHandle] !== undefined) {
+        this[kHandle].stats[IDX_QUIC_STREAM_STATS_DESTROYED_AT] =
+          process.hrtime.bigint();
+        state.stats = new BigInt64Array(this[kHandle].stats);
+      }
+      state.sharedState = undefined;
       if (state.dataRateHistogram)
         state.dataRateHistogram[kDestroyHistogram]();
       if (state.dataSizeHistogram)
         state.dataSizeHistogram[kDestroyHistogram]();
       if (state.dataAckHistogram)
         state.dataAckHistogram[kDestroyHistogram]();
     }
+    this[kHandle] = handle;
   }
 
   [kStreamReset](code) {
-    this[kInternalState].resetCode = code | 0;
+    // Receiving a reset from the peer indicates that it is no
+    // longer sending any data, we can safely close the readable
+    // side of the Duplex here.
+    this[kInternalState].resetCode = code;
     this.push(null);
     this.read();
   }
 
+  [kClose]() {
+    const state = this[kInternalState];
+
+    if (this.destroyed) {
+      return PromiseReject(
+        new ERR_INVALID_STATE('QuicStream is already destroyed'));
+    }
+
+    const promise = deferredClosePromise(state);
+    if (this.readable) {
+      this.push(null);
+      this.read();
+    }
+
+    if (this.writable) {
+      this.end();
+    }
+
+    return promise;
+  }
+
+  close() {
+    return this[kInternalState].closePromise || this[kClose]();
+  }
+
+  _destroy(error, callback) {
+    const state = this[kInternalState];
+    const handle = this[kHandle];
+    this[kSetHandle]();
+    if (handle !== undefined)
+      handle.destroy();
+    state.session[kRemoveStream](this);
+
+    if (error && typeof state.closePromiseReject === 'function')
+      state.closePromiseReject(error);
+    else if (typeof state.closePromiseResolve === 'function')
+      state.closePromiseResolve();
+
+    process.nextTick(() => callback(error));
+  }
+
+  [kDestroy](code) {
+    // TODO(@jasnell): If code is non-zero, and stream is not otherwise
+    // naturally shutdown, then we should destroy with an error.
+
+    // At this point, the underlying stream at the ngtcp2/c++ level
+    // is closed and is no longer usable. Any outbound writes still
+    // pending in the buffer need to be canceled because there is
+    // no longer anything to write them to and the writable side
+    // must be shutdown.
+    const writableState = this._writableState;
+    for (let n = writableState.bufferedIndex;
+      n < writableState.buffered.length; ++n) {
+      const { chunk, callback } = writableState.buffered[n];
+      const len = writableState.objectMode ? 1 : chunk.length;
+      writableState.length -= len;
+      // TODO(@jasnell): Proper error
+      callback(new ERR_OPERATION_FAILED('QuicStream was closed'));
+    }
+    this.end();
```

Comment:
Why not just `this.destroy()` which will error all the callbacks.

Reply:
Because I've been trying to track down a timing bug that was making that not work ;-) ... I think I have it tho and *should* be able to make this a call to `destroy()` immediately after clearing the handle below.